### ISSUE TITLE
ASTF traffic duration is shorter than expected

### ISSUE THREAD

**Issue by jsmoon** [cisco-system-traffic-generator/trex-core#606]
Hi, I recently found that my ASTF traffic duration is much shorter than expected.

I have programmed to send UDP traffic 9 packets per second for 120 seconds.
The client sends the 3 packets and receives 3 packets at first. And then it sends 1080 packets for 120 seconds.
The server receives the first 3 packets and sends 3 packets. And then it waits for 120 seconds and receives all packets.
The send rate and waiting are implemented by the `delay()` command.
The send rate uses `delay(111100)` and `delay(111200)`. The waiting uses `delay(120*1000*1000)`.

As you can see in the following result, the client's traffic duration is less than expected.
```
     m_traffic_duration |        116.12 sec |           120 sec | measured traffic duration 
                      - |                   |                   |                         
                    TCP |                   |                   |                         
                      - |                   |                   |                         
                      - |                   |                   |                         
                    UDP |                   |                   |                         
                      - |                   |                   |                         
           udps_accepts |                 0 |                 1 | connections accepted    
          udps_connects |                 1 |                 0 | connections established 
            udps_closed |                 1 |                 1 | conn. closed (includes drops) 
           udps_sndbyte |           1474596 |                36 | data bytes transmitted  
            udps_sndpkt |              1083 |                 3 | data packets transmitted 
           udps_rcvbyte |                36 |           1474596 | data bytes received     
            udps_rcvpkt |                 3 |              1083 | data packets received   
                      - |                   |                   |                         
```
I think this error rate (over 3.23%) is **quite large for my traffic simulation**.

According to my analysis, this issue is caused by the level 1 timer tick unit.
The level 0 timer tick unit is 0.02 msec. It covers 0 ~ 40.96 msec by 2048 wheel size.
And the level 1 timer tick unit is 5.12 msec. It means that the maximum error rate could be 12.5%.

I need a more precise action for the `delay()` timer.
I think there are 2 approaches to solve this issue.
1. lower the level 1 timer error rate using the larger `level1_div` value. (8 -> 16, 32, 64, ...)
2. remove the level 1 timer error rate by starting the level 0 timer with `m_ticks_left`.

**I prefer solution 2** because solution 1 still has quite a large error rate.
But the solution 2 will cause one more timer action. The most timer callbacks will be performed by the level 0 timer.

@hhaim, what do you think about this issue? Do you have any idea to solve this issue?

**Comment by hhaim** [comment]
@jsmoon I wouldn't touch this. This is the heart of the engine and there are many considerations for setting the ratios like that. 
12.5% is  good enough for simulation of application traffic.

**Comment by jsmoon** [comment]
@hhaim, I have tried solution 2. This issue seems solved from the result.
But, I think it needs more discussion for the performance or other possible issues.
```
     m_traffic_duration |           120 sec |           120 sec | measured traffic duration 
                      - |                   |                   |                         
                    TCP |                   |                   |                         
                      - |                   |                   |                         
                      - |                   |                   |                         
                    UDP |                   |                   |                         
                      - |                   |                   |                         
           udps_accepts |                 0 |                 1 | connections accepted    
          udps_connects |                 1 |                 0 | connections established 
            udps_closed |                 1 |                 1 | conn. closed (includes drops) 
           udps_sndbyte |           1474596 |                36 | data bytes transmitted  
            udps_sndpkt |              1083 |                 3 | data packets transmitted 
           udps_rcvbyte |                36 |           1474596 | data bytes received     
            udps_rcvpkt |                 3 |              1083 | data packets received   
                      - |                   |                   |                         
```
Here is my change for this result.
```
diff --git a/src/h_timer.cpp b/src/h_timer.cpp
index 4fce23b..e81d337 100644
--- a/src/h_timer.cpp
+++ b/src/h_timer.cpp
@@ -359,11 +359,17 @@ uint16_t CNATimerWheel::on_tick_level_count(int level,
         if (!event) {
             break;
         }
-        if (event->m_ticks_left==0) {
+        if (event->m_ticks_left <= m_cnt_state) { // aware spread effect
             m_total_events--;
             cb(userdata,event);
         }else{
-            timer_start_rest(event,event->m_ticks_left);
+            event->m_ticks_left -= m_cnt_state;
+            if (event->m_ticks_left < m_wheel_size) { // use level 0 timer
+                m_total_events--;
+                timer_start(event, event->m_ticks_left - 1); // level 0 already proceed 1 tick
+            } else {
+                timer_start_rest(event, event->m_ticks_left);
+            }
         }
         cnt++;
         if ( cnt==m_cnt_per_iter)  {
@@ -392,7 +398,7 @@ na_htw_state_num_t CNATimerWheel::on_tick_level1(void *userdata,htw_on_tick_cb_t
         if (!event) {
             break;
         }
-        if (event->m_ticks_left==0) {
+        if (event->m_ticks_left < (1<<m_wheel_level1_shift)) {  // m_cnt_div
             m_total_events--;
             cb(userdata,event);
         }else{
@@ -442,14 +448,14 @@ RC_HTW_t CNATimerWheel::timer_stop (CHTimerObj *tmr){
 RC_HTW_t CNATimerWheel::timer_start_rest(CHTimerObj  *tmr, 
                                         htw_ticks_t  ticks){
 
-    htw_ticks_t nticks = (ticks+m_wheel_level1_err)>>m_wheel_level1_shift; 
+    htw_ticks_t nticks = ticks>>m_wheel_level1_shift;
     if (nticks<m_wheel_size) {
-        if (nticks<2) {
-            nticks=2; /* not on the same bucket*/
+        if (nticks<1) {
+            nticks=1; /* not on the same bucket*/
         }
-        tmr->m_ticks_left=0;
+        tmr->m_ticks_left = ticks - (nticks<<m_wheel_level1_shift);
         tmr->m_wheel=1;
-        m_timer_w[1].timer_start(tmr,nticks-1);
+        m_timer_w[1].timer_start(tmr,nticks);
     }else{
         tmr->m_ticks_left = ticks - ((m_wheel_size-1)<<m_wheel_level1_shift);
         tmr->m_wheel=1;
diff --git a/src/h_timer.h b/src/h_timer.h
index 0bef4f6..bca1a5a 100644
--- a/src/h_timer.h
+++ b/src/h_timer.h
@@ -497,7 +497,7 @@ public:
             tmr->m_wheel=0;
             return (m_timer_w[0].timer_start(tmr,ticks));
         }
-        return ( timer_start_rest(tmr, ticks));
+        return ( timer_start_rest(tmr, ticks + m_cnt_state)); // adjust level 1 timer tick offset
     }
 
     RC_HTW_t timer_stop (CHTimerObj *tmr);
```
I saved remain ticks to `m_ticks_left` at `timer_start_rest()` and it is used at `on_tick_level_count()` for `TCP_TW` and `on_tick_level1()` for `TW_SYNC1`.
I have considered the level 1 timer spread effect at `TCP_TW` for the precise calculation.
Since `TW_SYNC1` is separated from `TW_SYNC`, I didn't use the level 0 timer in this case.
And I have changed the `nticks` calculation from ceiling algorithm to flooring algorithm to get the left ticks easy.

@hhaim, could you review my trial changes?

**Comment by hhaim** [comment]
@jsmoon as I said this might look to you as a solution for this **simple** use case, but it will fail badly for the main use-cases (scale of flows).  Many factors were considered when setting the values. I just can say that this is not simple as it seems. 
The only option to _solve_ this is to add more levels (maybe 4) with a factor of 8 instead of 16 but this will give 7% and it will affect the performance in the cost of timer accuracy (which is acceptable).

**Comment by jsmoon** [comment]
@hhaim I agree that this solution is not simple. So I have tried to minimize the effect of this change.
The only added action is launching a new level 0 timer for the left ticks instead of the direct callback.
So, I thought the performance issue only. I couldn't understand why this action affects the scale of flows badly.
Could you elaborate on it for me?

**Comment by hhaim** [comment]
@jsmoon I've looked into the code, I changed to factors from 1024,16 (more accurate)  ~1.28 msec tick (level1) which in the worst case can have  6% to 2048/8 (12.5% accuracy in worst case)  for having more range to scale use-cases. So changing the factor to 32 will give  3% and having more levels will keep this accuracy with less buckets. 

Once you are at level 1 (or level x) the revolution is different (5msec), you can't get back to the level-0. The 12.5% is for the **worst case** for best case it could be 0.05% which is negligible (depends what is the timerout)

In short, I think that this issue is less of importance for the general case to give more dynamic range and having 4 levels of 1024 buckets with 32 drives would be a better solution.

**Comment by jsmoon** [comment]
@hhaim I'd like to know just what I don't understand. How this action affects the scale of flows?

> Once you are at level 1 (or level x) the revolution is different (5msec), you can't get back to the level-0.

I think I can go back to the level-0 by adjusting the `ticks` value with the synched value `m_cnt_state` at `timer_start()`.
```
-        return ( timer_start_rest(tmr, ticks));
+        return ( timer_start_rest(tmr, ticks + m_cnt_state)); // adjust level 1 timer tick offset
```
My only concern was the performance issue by adding a new level 0 timer instead of the timeout callback.

> having 4 levels of 1024 buckets with 32 drives would be a better solution.

Do you mean that adds level 2 and level 3 timer? I think this is quite complicated.

**Comment by hhaim** [comment]
@jsmoon you are pushing me to explain more details .. 
In your case the error is 3% because of the timer you set (11,000usec), which is ~5000 ticks, higher than 2048 to  level1.  In case of higher value the error will be lower (as low as 0.05%)
In case you will set it to 50,000usec you will get ~12% error. This is the design. 
The rationale behind  it is this: if you set a timer of 5 sec you don't need 20usec accuracy which is 0.0004% accurate. 
The issue is with a lower timer that just breached the lower level and pass  a higher level (in the past worst case of 6% now 12%)

In case of 32 div, you will get on your case **0.7%** error is it enough? 

Adding another timer callback from level1 to level 0 is missing the point, it is not only not needed when you set the ratio correctly, it would have a **significant ** (at least factor 2 worse) scale issue in case of 50-40M active flows because most of the CPU in this case is working on handle the timer (ack timer and keep alive)

I don't think that adding more levels is more complex, this is the right way to go. This will give tradeoff of dynamic range and error in percent. I agree that this is a sensitive piece of code and affect the performance, so need to be careful. 
I will look

**Comment by hhaim** [comment]
@jsmoon 
Have a look into this branch [timer_w_b](https://github.com/hhaim/trex-core/tree/timer_h), havn't test it yet. 
if you change the 16 to 32 you will get ~0.7%

**Comment by jsmoon** [comment]
@hhaim I think now I understood what you are worrying about.
The timer `CNATimerWheel` is used for TCP layer timer mainly.
And I agree that it doesn't need to be too accurate at the expense of performance.
So, my preferred 'solution 2' was not a good approach.

Anyway, this issue was the `delay()` command accuracy.
Since it is used for the app running, I think it needs a more accurate operation.
If there is a 1% error rate for the delay, for the 120 sec period in my case, it will be finished 1 second earlier.
User requests the traffic for 120 sec on a flow, my app runs 119 sec. I'd like to avoid this situation.
I'll find another solution without changing the TCP layer timer.

**Comment by hhaim** [comment]
@jsmoon I don't understand why 0.7% (almost worst case) Is good for TCP but does not good for application layer? if it has been 100 hours as a delay why would you want 20usec as a resolution who care? 
In both cases the timer does not need to be accurate in the same way and it is good that there is randomness (this is the reason we added this command) 

There is another **simple** solution that can make the error close to zero and can be exposed only to the application layer (still not sure it is needed but it is simple) 

In your example, it will give 120 sec in almost zero cost. The idea is to calculate the error and add one in the probability of the range. Example:
Let's say you you need 16.1 ticks, but choose 16 (error 0.1/16) each time you ask 16.1 you throw a random coin, 1 out of 10 you will choose 17 instead of 16. In many timer starts it will give exact accumulation 
this could be exposed only as a command to app layer

**Comment by jsmoon** [comment]
@hhaim, thank you for your review of this issue.

> I don't understand why 0.7% (almost worst case) Is good for TCP but does not good for application layer? if it has been 100 hours as a delay why would you want 20usec as a resolution who care?

As I mentioned in the previous comment,
>> User requests the traffic for 120 sec on a flow, my app runs 119 sec. I'd like to avoid this situation.

In this case, my user monitors my app running and complain about why it runs less than requested.
If the duration is 1200 sec, the user time difference can be up to 10 seconds. The time gap will matter to my user.

If there is another way to implement the required traffic by the app (without using the `delay` command), I think it will be a better solution for this issue.

> There is another simple solution that can make the error close to zero and can be exposed only to the application layer (still not sure it is needed but it is simple)

I'll prepare a PR for this suggestion next year.

> In your example, it will give 120 sec in almost zero cost. The idea is to calculate the error and add one in the probability of the range. Example:
Let's say you you need 16.1 ticks, but choose 16 (error 0.1/16) each time you ask 16.1 you throw a random coin, 1 out of 10 you will choose 17 instead of 16. In many timer starts it will give exact accumulation
this could be exposed only as a command to app layer

My app had been used the error calculation already for the level 0 timer tick(20usec). But, for the level 1 timer tick, I didn't realize the error rate before. The calculation was not easy for me when I tried to make it more accurate.

**Comment by hhaim** [comment]
@jsmoon I’m going to commit my side branch with 4 levels. In this case your 120 sec will have 0.2% error which is ~200 msec. In case of 240 second it will be 0.1%.

You can verify it.

My suggestion will work on accumulation of many timers giving on average exactly the requested value. It will not help with a single timer.

**Comment by jsmoon** [comment]
@hhaim, your branch commit may affect my `delay` command usage for the traffic duration.

According to my calculation with 1024 wheel size and 16 div value,
Level | Time range | Tick unit | Error rate (max.)
--- | --- | --- | ---
0 | ~ 20.48 msec | 20 usec | 0.098% ~
1 | ~ 1.31 sec | 1.28 msec | 0.098% ~ 6.25%
2 | ~ 83.89 sec | 81.92 msec | 0.098% ~ 6.25%
3 | 83.89 sec ~ | 5.24 sec | ~ 6.25%

For the single delay of 83.89 sec, the time gap may be up to 5.24 seconds.
So, I think the `delay` change should be done also.

Thank you.

**Comment by hhaim** [comment]
@jsmoon your table is accurate. At all levels the error is betwean (100/16 (~6%) - 100/1024 = ~0.1%)  even at the last level. 
delay(111100) => 86 ticks => 100/86 = 1.6% error. The accumulation trick will reduce it to almost zero. 
Please implement it on top of my branch  ( add a timer.start (accurate = false) ) -- I think the random should be added only if the ticks is less that let say 500 (100/500 = 0.2%) and not at all cases

**Comment by hhaim** [comment]
@jsmoon 
I've pushed the new code, you should add a random here, based on an input (enable_accurate) 
https://github.com/cisco-system-traffic-generator/trex-core/blob/master/src/h_timer.cpp#L461
I would start by adding it by default (default is enabled),  If the perf impact is low, we could add it by default (without adding ASTF instruction to enable this)

**Comment by hhaim** [comment]
@jsmoon could you check this code
https://github.com/hhaim/trex-core/tree/test_t1

I ran a simple simulation and it seems to works (1000 ticks of 5555) 

the average is right, 
```
action-id: 997 tick : 5594  
avr : 5555  
```

please verify.

There is no need to use this ```m_exd_timer[m_max_levels-1].m_cnt_state```  as you can't do the reverse spreading calculation 
and there is no need to change the behaviors for the classic use-case (2 levels), this way the old STF golden files is the same.

**Comment by jsmoon** [comment]
@hhaim I've done it with my simulation. I agreed that it works as expected.

I added `m_exd_timer[m_max_levels-1].m_cnt_state` just for more precise calculation. In your random approach, I agree that there is no issue without it.

So, I think I'd better abandon my change and you push your branch.

**Comment by jsmoon** [comment]
@hhaim I gave up my PR #607. When do you merge your branch?

**Comment by hhaim** [comment]
@jsmoon ok, I will try to add the accurate capability too

**Comment by hhaim** [comment]
@jsmoon it was commited
